worker/objectstore/worker.go

```diff
@@ -0,0 +1,325 @@
+// instances into the worker loop.
+type objectStoreRequest struct {
+	namespace string
+	done      chan error
+}
+
+type objectStoreWorker struct {
+	internalStates chan string
+	cfg            WorkerConfig
+	catacomb       catacomb.Catacomb
+
+	runner *worker.Runner
+
+	objectStoreRequests chan objectStoreRequest
+}
+
+// NewWorker creates a new object store worker.
+func NewWorker(cfg WorkerConfig) (*objectStoreWorker, error) {
+	return newWorker(cfg, nil)
+}
+
+func newWorker(cfg WorkerConfig, internalStates chan string) (*objectStoreWorker, error) {
+	var err error
+	if err = cfg.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &objectStoreWorker{
+		internalStates: internalStates,
+		cfg:            cfg,
+		runner: worker.NewRunner(worker.RunnerParams{
+			Clock: cfg.Clock,
+			IsFatal: func(err error) bool {
+				return false
+			},
+			RestartDelay: time.Second * 10,
+			Logger:       cfg.Logger,
+		}),
+		objectStoreRequests: make(chan objectStoreRequest),
+	}
+
+	if err = catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{
+			w.runner,
+		},
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return w, nil
+}
+
+func (w *objectStoreWorker) loop() (err error) {
+	// Report the initial started state.
+	w.reportInternalState(stateStarted)
+
+	for {
+		select {
+		// The following ensures that all objectStoreRequests are serialised and
+		// processed in order.
+		case req := <-w.objectStoreRequests:
+			if err := w.initObjectStore(req.namespace); err != nil {
+				select {
+				case req.done <- errors.Trace(err):
+				case <-w.catacomb.Dying():
+					return w.catacomb.ErrDying()
+				}
+				continue
+			}
+
+			select {
+			case req.done <- nil:
+			case <-w.catacomb.Dying():
+				return w.catacomb.ErrDying()
+			}
+
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		}
+	}
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *objectStoreWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *objectStoreWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+// GetObjectStore returns a objectStore for the given namespace.
+func (w *objectStoreWorker) GetObjectStore(ctx context.Context, namespace string) (coreobjectstore.ObjectStore, error) {
+	// First check if we've already got the objectStore worker already running.
+	// If we have, then return out quickly. The objectStoreRunner is the cache,
+	// so there is no need to have a in-memory cache here.
+	if objectStore, err := w.workerFromCache(namespace); err != nil {
+		if errors.Is(err, w.catacomb.ErrDying()) {
+			return nil, coreobjectstore.ErrObjectStoreDying
+		}
+
+		return nil, errors.Trace(err)
+	} else if objectStore != nil {
+		return objectStore, nil
+	}
+
+	// Enqueue the request as it's either starting up and we need to wait longer
+	// or it's not running and we need to start it.
+	req := objectStoreRequest{
```

Comment:
This is confusing. As I understand this is just a lock until we have `initObjectStore`'d the object store. This worker's loop is only doing that.
So can't we just use a `sync.Mutex` and return a cached object store if it's indeed already init'd? By the way, do we need this as a worker then?

Reply:
> So can't we just use a sync.Mutex and return a cached object store if it's indeed already init'd

Nope. There has been a lot of work put into ensuring that we correctly play nice with the runner package. There have been a lot of work put into this pattern to ensure that we correctly instantiate the worker in the runner.

You shouldn't create the worker outside of the runner (we do in other places and that needs fixing) and so you have to serialize the work. This ensures the correct guarding of potential races. Two workers were created, but only one registered with the runner. Thus leaking a worker.